Trigger a raft election after splits

bdarnell opened the pull request:
This minimizes the window of unavailability following a split.

Fixes #1384.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3160)

<!-- Reviewable:end -->

bdarnell committed:
Trigger a raft election after splits.

This minimizes the window of unavailability following a split.

Fixes #1384.

tschottdorf commented:
LGTM. The test seems to be stuck in an endless loop related to elections, though.

---

Reviewed 1 of 1 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

_[storage/replica_command.go, line 1371 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3160#-K3PjAGtrY1_WyQ0y3Rm:-K3PjAGtrY1_WyQ0y3Rn:2095990434) ([raw file](https://github.com/cockroachdb/cockroach/blob/e979cabe22dad3e5a486493308a6debb3f6dbdb0/storage/replica_command.go#L1371)):_
Seems worthwhile implementing the above suggestion right away.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3160#-:-K3Pjjj6gkkicD6Ywzvu:993928084)_
<!-- Sent from Reviewable.io -->

es-chow commented on storage/replica_command.go:
```diff
@@ -1359,6 +1359,18 @@ func (r *Replica) splitTrigger(batch engine.Engine, split *roachpb.SplitTrigger)
+		// improve this by e.g. choosing the node that had the leader
+		// lease before the split and is therefore known to be up)
+		if r.store.StoreID() == split.NewDesc.Replicas[0].StoreID {
+			r.store.multiraft.Campaign(split.NewDesc.RangeID)
```
If splitting of the replicas on other node are not finished, then this campaign is no-op, the group still have to wait for an election timeout.
In our test, we add 10ms delay before start the campaign, it will improve the chance of success.

bdarnell replied:
That shouldn't be necessary. The `storage.Replica` can't be created until after the other nodes have processed the split, but the multiraft group is created (as an uninitialized replica) in response to incoming messages and can still participate in elections.

bdarnell replied:
Ah, but there is still a problem here. An uninitialized replica can vote, but when the replica becomes initialized the raft group is deleted and recreated, discarding any pending messages. It's a rare race (for it to matter *both* followers need to drop their vote messages), so it might be OK to just wait for an election timeout when it happens, but in this new test that's not an option and so the test fails about 1 in 200 times.

bdarnell replied:
As long as we drop and recreate the raft group, I don't see any way around this. It's not a huge problem in practice (because the worst case is just waiting for an election timeout), but the failure is frequent enough that the test will be flaky. I'm going to explore an alternate solution: instead of dropping the uninitialized group and recreating it as an initialized one, synthesize a snapshot (containing only metadata) to send to ourselves and initialize the group that way. If that works, it should make the test robust. If not, I'll have to fall back to either adding a 10ms delay (if that works consistently) or running the test multiple times and ensure that the election succeeds a majority of the time.

bdarnell replied:
The second commit in this branch attempts to address this issue. It avoids the problem of dropping and recreating the raft group, but it creates another similar problem because the range metadata (replicasByKey) in the Store is no longer updated atomically. I think this could be solved with some more work, but I'm not sure it's worth it, so I'm probably going to back to hacking around the problem just enough to make this test reliable.

bdarnell committed:
Add a small delay before campaigning after split.

This improves the robustness of TestLeaderAfterSplit.

bdarnell committed:
The node that started a split should become leader after it.

bdarnell commented:
Updated the branch to add a small delay before campaigning per @es-chow 's suggestion. The abandoned commit I mentioned in my earlier comment has been preserved in https://github.com/bdarnell/cockroach/tree/init-split-by-snapshot. I also implemented @tschottdorf 's suggestion to make a better choice of initial leader.

tschottdorf commented:
LGTM mod comments.

---

Reviewed 4 of 4 files at r3, 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/replica_command.go, line 1438 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3160#-K4OeCPnDJKL645cX341:-K4OeCPnDJKL645cX342:-1405202666) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbc9b280143bb83bf3d7c0226535517205ad8b8c/storage/replica_command.go#L1438)):*
while the fact that you're putting a delay in here makes pretty clear that there's barely a way around it, can we turn that into a package-level constant that's usually 0 (omitting the call `Sleep`) and only set it in the test that would otherwise be flaky?

---

*[storage/replica_command.go, line 1318 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3160#-K4Of-0Wwuui65-Qhmt1:-K4Of-0Wwuui65-Qhmt2:-1446974008) ([raw file](https://github.com/cockroachdb/cockroach/blob/9034022d8443597fde72ed05a19fb359d0e446d9/storage/replica_command.go#L1318)):*
I thought the main reason (on top of yours) for doing this was that this would also be the node that has the most direct r/w traffic.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3160#-:-K4OfGLozLwxzl_j0NOJ:1016183029)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/replica_command.go, line 1438 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3160#-K4OeCPnDJKL645cX341:-K4T0KMPQCYjEZBBfgza:1944885685) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbc9b280143bb83bf3d7c0226535517205ad8b8c/storage/replica_command.go#L1438)):*
We could, but I'm concerned about the test diverging from the real world if we did that. Previously, es-chow found that the delay was necessary in all cases. That's no longer true but we have no reliable way of testing without a delay. If we had a regression on this issue we would have a situation where the test passed because it inserted a delay, but it wasn't working in the real world. 

---

*[storage/replica_command.go, line 1318 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3160#-K4Of-0Wwuui65-Qhmt1:-K4SzERn8lzQTvFOtFDB:-805707004) ([raw file](https://github.com/cockroachdb/cockroach/blob/9034022d8443597fde72ed05a19fb359d0e446d9/storage/replica_command.go#L1318)):*
In the long run that should be true, but currently our leader leases are only weakly correlated with traffic (and even if the leader was the node with the most traffic, in deployments with small geographic scale it may be better to balance the load than to choose the same node as leader). My main concern in making this change is to avoid choosing a down node which would require waiting for an election timeout.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3160)*
<!-- Sent from Reviewable.io -->
